### PR TITLE
interfaces: add consoles interface

### PR DESCRIPTION
**Pull request by femdom**
To allow access to the /dev/tty0 and /dev/console

### CODE REVIEW

**Comment by ogra1** [comment]
i dont think this will fly as is, most ubuntu-core installs will actually use a different console= option to run their IoT devices with serial console etc ... the apparmor managing code should better parse /proc/cmdline and grab the console= arg from there instead of hardcoding tty0

**Comment by femdom** [comment]
@niemeyer 

>  Also, can you please provide some more background so we can understand the demand for this interface, can take the appropriate action regarding what @ogra1 is pointing out?

Our software tries to access /dev/tty0 or /dev/console or /dev/tty [qlinuxfbscreen.cpp](https://github.com/qt/qtbase/blob/5.6/src/plugins/platforms/linuxfb/qlinuxfbscreen.cpp#L243)

**Comment by stolowski** [comment]
@femdom FYI, I've updated this PR for the new API changes in master.

**Comment by femdom** [comment]
@stolowski, thank you!

File: interfaces/builtin/consoles.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/sys/devices/virtual/tty/console rw,
+`
+
+type ConsolesInterface struct{}
+
+func (iface *ConsolesInterface) Name() string {
+	return "consoles"
+}
+
+func (iface *ConsolesInterface) String() string {
```
Comment (zyga) [inline interfaces/builtin/consoles.go]: I don't think we need the `String` method really

File: interfaces/builtin/consoles.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func (iface *ConsolesInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	spec.AddSnippet(consolesConnectedPlugAppArmor)
+	return nil
+}
+
+func (iface *ConsolesInterface) UdevConnectedPlug(spec *udev.Specification, plug *interfaces.Plug, slot *interfaces.Slot) error {
+	for appName := range plug.Apps {
+		tag := udevSnapSecurityName(plug.Snap.Name(), appName)
```
Comment (zyga) [inline interfaces/builtin/consoles.go]: This begs to be one function but this is not a material for this PR

File: interfaces/builtin/consoles.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// Check validity of the defined slot
+func (iface *ConsolesInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget or os snap
+	if !(slot.Snap.Type == "os") {
```
Comment (mvo5) [inline interfaces/builtin/consoles.go]: I think we want `snap.TypeOS` here instead of hardcoding the string.

**Comment by jdstrand** [comment]
FYI, this interface is on my list for review but it will take a little while to get through everything.

**Comment by femdom** [comment]
Should I fix addressed issues?

**Comment by jdstrand** [comment]
Please feel free to address any review feedback, yes. It is still on my list to review and still will take quite a bit of time to get through everything. It has unfortunately fallen behind a number of other items, but we'll get there.

**Comment by niemeyer** [comment]
@jdstrand Any news here?  We're making a push to get the queue clean this week:

https://forum.snapcraft.io/t/review-sprint-1/377

**Comment by jdstrand** [comment]
@niemeyer - I'd like to study the tty subsystem for this PR to better understand it. I mentioned yesterday that this PR has fallen behind other work-- that work is mostly items from https://forum.snapcraft.io/t/review-sprint-1/377 (atm, bash completion, but several others after that).

This PR won't be reviewed this week at this point.

File: interfaces/builtin/consoles.go (hunk 3 above)
Comment (zyga) [inline interfaces/builtin/consoles.go]: Nitpick: can this please say `if slot.Snap.Type != snap.TypeOS { ... }` please?

**Comment by zyga** [comment]
I'll work on this first thing tomorrow. I'll review it again, apply all the feedback and get it ready for another review.

**Comment by zyga** [comment]
After discussing with @jdstrand I'm marking this interface as blocked. It requires additional research into the kernel tty layers to determine if the security is sensible.

**Comment by chipaca** [comment]
What's the state of this PR?

**Comment by jdstrand** [comment]
@chipaca - still blocked behind other higher priority work, but it is in the queue and not lost.

**Comment by chipaca** [comment]
OK. It'll need some serious de-conflicting I fear.

**Comment by ogra1** [comment]
can we please also have the sys_tty_config capability, something taking over the console might want to reconfigure the tty as well...

**Comment by zyga** [comment]
I just de-conflicted it, I will iterate some more to add modern meta-data and simplify a few things.

**Comment by jdstrand** [comment]
Since this continues to wait on me, I'm happy to deconflict it as needed (though I see @zyga did this already). For now I'm going to close this PR but will reopen when it is at the top of the queue.

**Comment by zyga** [comment]
I'll re-open this briefly to push my changes :)

**Comment by zyga** [comment]
Pushed, closing again :)

**Pull request closed without merging**